src/components/queryBuilder/EditorTypeSwitcher.tsx

```diff
@@ -0,0 +1,108 @@
+import React, { useState } from 'react';
+import { SelectableValue } from '@grafana/data';
+import { RadioButtonGroup, ConfirmModal, InlineFormLabel } from '@grafana/ui';
+import { getQueryOptionsFromSql } from '../queryBuilder/utils';
+import { generateSql } from 'data/sqlGenerator';
+import labels from 'labels';
+import { EditorType, CHQuery, defaultCHBuilderQuery } from 'types/sql';
+import { QueryBuilderOptions } from 'types/queryBuilder';
+import isString from 'lodash/isString';
+import { mapQueryTypeToGrafanaFormat } from 'data/utils';
+
+interface CHEditorTypeSwitcherProps {
+  query: CHQuery;
+  onChange: (query: CHQuery) => void;
+  onRunQuery: () => void;
+}
+
+const options: Array<SelectableValue<EditorType>> = [
+  { label: labels.types.EditorType.sql, value: EditorType.SQL },
+  { label: labels.types.EditorType.builder, value: EditorType.Builder },
+];
+
+/**
+ * Component for switching between the SQL and Query Builder editors.
+ */
+export const EditorTypeSwitcher = (props: CHEditorTypeSwitcherProps) => {
+  const { query, onChange } = props;
+  const { label, tooltip, switcher, cannotConvert } = labels.components.EditorTypeSwitcher;
+  const editorType: EditorType = query.editorType || EditorType.Builder;
+  const [confirmModalState, setConfirmModalState] = useState<boolean>(false);
+  const [cannotConvertModalState, setCannotConvertModalState] = useState<boolean>(false);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const onEditorTypeChange = (editorType: EditorType, confirm = false) => {
+    // TODO: component state has updated, but not local state.
+    if (query.editorType === EditorType.SQL && editorType === EditorType.Builder && !confirm) {
```

Comment:
Does `confirm` mean "require confirmation" or "confirmed switching"? let's rename to show the intention